ouroboros-consensus/src/Ouroboros/Consensus/Node/DbLock.hs

```diff
@@ -0,0 +1,47 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Ouroboros.Consensus.Node.DbLock (
+    DbLocked (..)
+  , withLockDB
+  , dbLockFile
+  ) where
+
+import           Control.Monad.Class.MonadAsync
+import           Control.Monad.Class.MonadTimer
+import           Data.Text
+import qualified Data.Time.Clock as Time
+import           System.FileLock
+
+import           Ouroboros.Consensus.Storage.FS.API
+import           Ouroboros.Consensus.Storage.FS.API.Types
+import           Ouroboros.Consensus.Util.IOLike
+
+-- We use an empty file as a lock of the db. Some systems may delete the empty
+-- file when all its handles are closed. This is not an issue, since the file is
+-- created if it doesn't exist.
+withLockDB :: HasFS IO h ->  FilePath -> IO a -> IO a
+withLockDB hasFS dbPath action = do
+    createDirectoryIfMissing hasFS True root
+    -- We want to avoid blocking the main thread at an ffi, to avoid
+    -- unresponsiveness. We shouldn't be tempted to use `withAsync` or anything
+    -- else which links the blocked thread to our main thread, since this would
+    -- negate the whole point of using async.
+    a     <- async (lockFile lockFilePath Exclusive)
+    mlock <- timeout (Time.secondsToDiffTime 2) $ wait a
+    case mlock of
+      Nothing   -> throwM $ DbLocked lockFilePath
+      Just lock -> finally action (unlockFile lock)
```

Comment:
There's some `mask`ing missing. The simplest approach is to write a `acquireLock` and a `releaseLock` function and use those in `bracket`, then `bracket` will take care of the masking for us.

Reply:
@mrBliss what do you think should be masked?

Reply:
What if an async exception is thrown after obtaining the lock and before `finally`? That's the thing with async exceptions, they can happen at any point.

Reply:
I'm not sure if we should mask while waiting on a timeout. Let me check it..